applications/HDF5Application/custom_io/hdf5_file.cpp

```diff
@@ -960,67 +990,77 @@ void File::WriteDataSetImpl(
+        bool write_data = TDataTransferMode == DataTransferMode::Collective;
 
-                // select the local hyperslab
-                KRATOS_HDF5_CALL(H5Sselect_hyperslab, fspace_id, H5S_SELECT_SET, local_shape_start.data(), nullptr, local_reduced_shape.data(), nullptr)
-                KRATOS_HDF5_CALL_WITH_RETURN(mspace_id, H5Screate_simple, global_dimension, local_reduced_shape.data(), nullptr)
-                if (number_of_local_primitive_data_values > 0) {
-                    KRATOS_HDF5_CALL(H5Dwrite, dset_id, dtype_id, mspace_id, fspace_id, dxpl_id, TypeTraits::GetContiguousData(rData))
-                } else {
-                    KRATOS_HDF5_CALL(H5Dwrite, dset_id, dtype_id, mspace_id, fspace_id, dxpl_id, nullptr)
-                }
+        // HDF5 methods for DataTransferMode::Independent should be only called from ranks where there is something to write.
+        write_data |= std::accumulate(local_shape.begin(), local_shape.end(), 1u, std::multiplies<hsize_t>{}) > 0;
 
-                KRATOS_HDF5_CALL(H5Pclose, dxpl_id)
-                KRATOS_HDF5_CALL(H5Sclose, mspace_id)
-            }
-        #else
-            KRATOS_ERROR << "HDFApplication is not compiled with MPI enabled";
+        #if H5_VERS_MAJOR < 2 && ((H5_VERS_MINOR == 14 && H5_VERS_RELEASE < 2) ||  H5_VERS_MINOR < 14)
+            /**
+             *  Until hdf5 1.14.2, if someone tries to read/write empty containers from every rank, it throws an error. This is
+             *  fixed in the later versions. This is valid for both collective and independent writing mechanisms.
+             *
+             * TODO: Remove this pre-compiler directive once we move to compatible versions.
+             */
+            write_data &= std::accumulate(global_shape.begin(), global_shape.end(), 1u, std::multiplies<hsize_t>{}) > 0;
         #endif
+
+        if (write_data) {
+            hid_t dxpl_id, mspace_id;
+            KRATOS_HDF5_CALL_WITH_RETURN(dxpl_id, H5Pcreate, H5P_DATASET_XFER)
+
+            if constexpr(TDataTransferMode == DataTransferMode::Collective) {
+                // if it is collective, calculate the offsets for each rank.
+                offset[0] = std::accumulate(
+                    r_collected_local_shapes.begin(),
+                    r_collected_local_shapes.begin() + r_data_communicator.Rank(),
+                    0u, [](const hsize_t Value, const auto& rRankShape) {
+                        return Value + rRankShape[0];
+                    });
+                KRATOS_HDF5_CALL(H5Pset_dxpl_mpio, dxpl_id, H5FD_MPIO_COLLECTIVE)
+            } else {
+                KRATOS_HDF5_CALL(H5Pset_dxpl_mpio, dxpl_id, H5FD_MPIO_INDEPENDENT)
+            }
+
+            // select the local hyperslab
+            KRATOS_HDF5_CALL(H5Sselect_hyperslab, fspace_id, H5S_SELECT_SET, offset.data(), nullptr, count.data(), nullptr)
+            KRATOS_HDF5_CALL_WITH_RETURN(mspace_id, H5Screate_simple, global_shape.size(), count.data(), nullptr)
```

Comment:
I guess this is semantically identical, but it makes me scratch my head less.

```suggestion
            KRATOS_HDF5_CALL_WITH_RETURN(mspace_id, H5Screate_simple, count.size(), count.data(), nullptr)
```